tensorrt_llm/_torch/models/modeling_llama.py

```diff
@@ -585,6 +594,72 @@ def load_weights(self, weights: Dict):
                     idx + 1].input_layernorm
 
 
+class Llama4InputProcessor(InputProcessor):
+    def __init__(self, model_path, model_config, tokenizer):
+        self.processor = AutoProcessor.from_pretrained(model_path, use_fast=True)
+        self.model_config = model_config
+        self.tokenizer = tokenizer
+        self.vocab_size = model_config.text_config.vocab_size
+        self.image_token_index = model_config.image_token_index
+
+        self.vision_model = Llama4VisionModel.from_pretrained(model_path).cuda()
+        self.multi_modal_projector = Llama4MultiModalProjector(model_config).cuda()
+        state_dict = self.multi_modal_projector.state_dict()
+        
+        with safe_open(Path(model_path) / "model.safetensors", framework="pt") as f:
+            for k in state_dict.keys():
+                state_dict[k] = f.get_tensor("multi_modal_projector." + k)
+
+        self.multi_modal_projector.load_state_dict(state_dict)
+
+    @torch.inference_mode()
+    def __call__(
+        self, inputs: TextPrompt, sampling_params: SamplingParams
+    ) -> Tuple[List[int], Optional[ExtraProcessedInputs]]:
+        text_prompt, mm_data = inputs.get("prompt"), inputs.get("multi_modal_data")
+        if not mm_data or not mm_data.get("image"):
+            # text only
+            return token_ids, {}
+        
+        # preprocess images and insert image tokens
+        img_type = type(mm_data["image"][0])
+        assert all(isinstance(img, img_type) for img in mm_data["image"])
+
+        processed = self.processor(
+            text=text_prompt, images=mm_data["image"],
+            return_tensors="pt", device="cuda",
+            do_rescale=(img_type == Image)
+        )
+        token_ids, pixel_values = processed["input_ids"].squeeze(), processed["pixel_values"]
+        mm_embeds = self.vision_model(pixel_values.float().cuda()).last_hidden_state.flatten(0, 1)
+        mm_embeds = self.multi_modal_projector(mm_embeds)
+        # for fuse_input_embeds
+        token_ids[token_ids == self.image_token_index] = self.vocab_size + 1
+        return token_ids.tolist(), {"prompt_tuning_config": [mm_embeds, None, None]}
+
+
+
+@register_auto_model("Llama4ForConditionalGeneration")
+@register_input_processor(Llama4InputProcessor)
+class Llama4ForConditionalGeneration(Llama4ForCausalLM):
+    @torch.inference_mode()
+    def forward(
+        self,
+        attn_metadata: AttentionMetadata,
+        input_ids: Optional[torch.LongTensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        return_context_logits: Optional[bool] = False,
+        **kwargs,
+    ) -> torch.Tensor:
+        mm_embed = kwargs.get("multi_modal_data", [])
+        input_ids, inputs_embeds = fuse_input_embeds(self, input_ids, mm_embed)
```

Comment:
should self be changed to embedding layer here?

Reply:
Yep